File: String Permutation/code.py


def toString(List): 
	return ''.join(List) 

def permute(a, s, l): 
	if s==l: 
		print toString(a) 
	else: 
		for i in xrange(s,l+1): 
			a[s], a[i] = a[i], a[s] 
			permute(a, s+1, l) 
			a[s], a[i] = a[i], a[s] # backtrack 

string = "ABC"
n = len(string) 
a = list(string) 
permute(a, 0, n-1) 

File: String Permutation/code.c
#include <stdio.h> 
#include <string.h> 

void swap(char *x, char *y) 
{ 
	char temp; 
	temp = *x; 
	*x = *y; 
	*y = temp; 
} 

/* Function to print permutations of string 
This function takes three parameters: 
1. String 
2. Starting index of the string 
3. Ending index of the string. */
void permute(char *a, int s, int l) 
{ 
int i; 
if (s == l) 
	printf("%s\n", a); 
else
{ 
	for (i = s; i <= l; i++) 
	{ 
		swap((a+s), (a+i)); 
		permute(a, s+1, l); 
		swap((a+s), (a+i)); //backtrack 
	} 
} 
} 

int main() 
{ 
	char str[] = "ABCDEF"; 
	int n = strlen(str); 
	permute(str, 0, n-1); 
	return 0; 
} 

